packages/core/src/modules/routing/protocol/pickup/v2/V2MessagePickupService.ts

```diff
@@ -0,0 +1,101 @@
+import { InjectionSymbols } from '../../../../../constants'
+import { Attachment } from '../../../../../decorators/attachment/Attachment'
+import { inject, injectable } from '../../../../../plugins'
+import { MessageRepository } from '../../../../../storage/MessageRepository'
+import { MediationRecipientService } from '../../../services'
+
+import {
+  DeliveryRequestHandler,
+  MessageDeliveryHandler,
+  MessagesReceivedHandler,
+  StatusHandler,
+  StatusRequestHandler,
+} from './handlers'
+import { MessageDeliveryMessage, StatusMessage } from './messages'
+
+@injectable()
+export class V2MessagePickupService {
+  private messageRepository: MessageRepository
+  private dispatcher: Dispatcher
+  private mediationRecipientService: MediationRecipientService
+
+  public constructor(
+    @inject(InjectionSymbols.MessageRepository) messageRepository: MessageRepository,
+    dispatcher: Dispatcher,
+    mediationRecipientService: MediationRecipientService
+  ) {
+    this.messageRepository = messageRepository
+    this.dispatcher = dispatcher
+    this.mediationRecipientService = mediationRecipientService
+
+    this.registerHandlers()
+  }
+
+  public async processStatus(messageContext: InboundMessageContext<StatusRequestMessage>) {
+    // Assert ready connection
+    const connection = messageContext.assertReadyConnection()
+
+    const statusMessage = new StatusMessage({
+      messageCount: await this.messageRepository.available(connection.id),
+    })
+
+    return createOutboundMessage(connection, statusMessage)
+  }
+
+  public async queueMessage(connectionId: string, message: EncryptedMessage) {
+    await this.messageRepository.add(connectionId, message)
+  }
+
+  public async processDeliveryRequest(messageContext: InboundMessageContext<DeliveryRequestMessage>) {
+    // Assert ready connection
+    const connection = messageContext.assertReadyConnection()
+
+    const { message } = messageContext
+    const messages = await this.messageRepository.takeFromQueue(connection.id, message.limit)
+
+    // TODO: each message should be stored with an id. to be able to conform to the id property
+    // of batch message
+    const attachments = messages.map(
+      (msg) =>
+        new Attachment({
+          data: {
+            json: msg,
+          },
+        })
+    )
+
+    const deliveryMessage = new MessageDeliveryMessage({
+      recipientKey: message.recipientKey,
+      attachments,
+    })
+
+    return createOutboundMessage(connection, deliveryMessage)
+  }
+
+  public async processMessagesReceived(messageContext: InboundMessageContext<MessagesReceivedMessage>) {
+    // Assert ready connection
+    const connection = messageContext.assertReadyConnection()
+
+    const { message } = messageContext
+
+    // TODO: Add Queued Message ID
+    await this.messageRepository.takeFromQueue(
+      connection.id,
+      message.messageIdList ? message.messageIdList.length : undefined
+    )
```

Comment:
We shouldn't take from the queue in this case I think? 

I'm assuming this is what you're talking about in the pr description. We should somehow be able to get messages from the queue (without removing them) and also be able to remove messages from the queue. Currently the acknowledgement doesn't matter and it will always just remove them right?

Reply:
Exactly. That's not yet handled properly. I was thinking on adding an optional parameter to `takeFromQueue` like `keepMessages` that defaults to current behaviour, so it does not break current V1 handler and limits changes to minimum for 0.2.x.

But if we are free to change `MessageRepository`/`InMemoryMessageRepository` and V1 service/handlers a bit (as long as `MediationModule` and `MediationRecipient` API does not break), probably it would be better to have a method to slice and another to splice the internal messages array.

Reply:
The message repo can already be registered with a custom version at the moment (I'm using it in a project) so we should be careful with changing it. 

I think for now this is fine and then we can go wild in the 0.3.0 branch. WDYT?

Reply:
Absolutely!